### DISCUSSION TITLE
Feature request: multi-reference ranges (eg. `A1:B2:C3:D4:E5`)

### DISCUSSION THREAD

**Discussion by argandsolutions-glyn** [handsontable/hyperformula#1460] [Feature requests]
**Idea**
Could we add support for individual cells within the sum/criteria ranges for SUMIFS?

**Example**
=SUMIFS(A2:B2:C2, D2:E2:F2, ">1") 

![Screenshot 2024-11-14 at 13 41 37](https://github.com/user-attachments/assets/70504451-38a3-4ba2-9b8c-36f36f281e71)

At present, i've found that only ranges (sum or criteria) within the same column are supported i.e. =SUMIFS(A2:A4, B2:B4, ">0")

I'd be happy to contribute towards this functionality if that would be helpful.

**Comment by sequba** [comment]
Hi @argandsolutions, thank you for reaching out.

I'm not sure what you mean. The example you provided works in HyperFormula if you remove the redundant middle cell addresses from the range specifications:

```javascript
import HyperFormula from 'hyperformula';

const hf = HyperFormula.buildFromArray([
  ['=SUMIFS(A2:C2, D2:F2, ">1")'],
  [1, 2, 3, 4, 5, 6]
],{
  licenseKey: 'gpl-v3',
});

console.log(hf.getSheetValues(0));
```

Can you describe the requested feature in more detail and provide more examples?

**Reply by argandsolutions-glyn** [reply]
Yes certainly.

What i'm looking to achieve is the following:

```
import HyperFormula from 'hyperformula';

const hf = HyperFormula.buildFromArray([
  ['=SUMIFS(A2:B2:C2, D2:E2:F2, ">1")'],
  [1, 2, 3, 4, 5, 6]
],{
  licenseKey: 'gpl-v3',
});

console.log(hf.getSheetValues(0));

// The above results in "Parsing error. Expecting token of type --> RParen <-- but found --> ':' <--"
```

This appears to be possible in Excel:

![image](https://github.com/user-attachments/assets/ec4e83f4-0770-49cc-9891-e21f6058cd4c)

At present, in my script I am building up the formula and dataset programatically and it would take further processing in order to reduce the formula inputs to a range (A2:C2) rather than individual cell references (A2:B2:C2). 

I hope that makes sense. If not, I can provide additional context and examples.

**Comment by sequba** [comment]
ODFF reference on ranges: https://docs.oasis-open.org/office/OpenDocument/v1.3/os/part4-formula/OpenDocument-v1.3-os-part4-formula.html#__RefHeading__1018032_715980110

**Reply by argandsolutions-glyn** [reply]
Ah ok. That's understandable.

I'd be happy to close this discussion if the feature i'm interested in isn't part of the OpenFormula format?

**Comment by sequba** [comment]
@argandsolutions It's a valid feature request. It just needs the proper specification. The HyperFormula team will probably consider it low-priority, but it's still worth discussing.

The main issue is that, as far as I can see, Microsoft Excel interprets such ranges differently from the OpenFormula standard.

### OpenFormula

According to [the Standard](https://docs.oasis-open.org/office/OpenDocument/v1.3/os/part4-formula/OpenDocument-v1.3-os-part4-formula.html#__RefHeading__1018032_715980110), range `A1:B2:C3:D4:E5` is equivalent to `A1:E5`.

![image](https://github.com/user-attachments/assets/3d7bc267-063d-4ae1-85bb-eb92830068ef)

### Microsoft Excel

I couldn't find any official reference describing the syntax of range specification in Excel, but it seems that `A1:B2:C3:D4:E5` is interpreted as the sum of 3 ranges:  `A1:B2` + `C3:D4` + `E5:E5`:

![image](https://github.com/user-attachments/assets/a4601be3-7686-403d-bb4e-c417c7f2784e)

If we were to consider adding this feature, we would need to choose one of the above semantics or implement both and introduce a configuration option to switch between them.

**Reply by argandsolutions-glyn** [reply]
Thanks for your thoughts @sequba 

As you say, it's difficult as the specification differs between the Excel implementation of the ranges used in the SUMIFS function and the OpenFormula standard. It then raises the question of whether HyperFormula is based on the OpenFormula standard or a hybrid of the two (Excel and OpenFormula), which might then cause confusion for users. As you mentioned before, you'd need a configuration option to allow users to switch between the two.

I wouldn't know how to proceed with this. It's essentially behaving fine if your expectations are based on OpenFormula but not if it's based on Excel.

Perhaps there's nothing we can do right now?

**Reply by sequba** [reply]
We stick to OpenFormula as much as possible while keeping in mind that most users expect the same behavior as in Excel. In this case, the ideal option is to make it configurable.

The Microsoft Excel interpretation is much more challenging to implement. It may require significant changes in the core of the engine. Also, it raises a question about how other functions should operate on such non-rectangular ranges (e.g. `TRANSPOSE`).

As I wrote above, either way, this feature is low-priority at the current stage of development, so we'll keep it as an idea for the future.

**Reply by argandsolutions-glyn** [reply]
That's understandable. It makes sense that it would be challenging to change how the core operates to accommodate Excel's behaviour.

Shall I close this discussion? Or does it need labelling to mark it as a future idea?

**Reply by sequba** [reply]
Let's keep it open in case someone else is also interested in this feature ;)

**Reply by argandsolutions-glyn** [reply]
Cool :)
